File: simplestore/src/main/java/com/uber/simplestore/primitive/PrimitiveSimpleStoreImpl.java
package com.uber.simplestore.primitive;

import static com.google.common.util.concurrent.MoreExecutors.directExecutor;

import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.uber.simplestore.SimpleStore;
import javax.annotation.Nullable;

@SuppressWarnings("UnstableApiUsage")
final class PrimitiveSimpleStoreImpl implements PrimitiveSimpleStore {

  private final SimpleStore simpleStore;

  PrimitiveSimpleStoreImpl(SimpleStore simpleStore) {
    this.simpleStore = simpleStore;
  }

  @Override
  public ListenableFuture<String> getString(String key) {
    return Futures.transform(
        simpleStore.getString(key), value -> value != null ? value : "", directExecutor());
  }

  @Override
  public ListenableFuture<String> put(String key, String value) {
    return simpleStore.putString(key, value);
  }

  @Override
  public ListenableFuture<String> putString(String key, @Nullable String value) {
    return simpleStore.putString(key, value);
  }

  @Override
  public ListenableFuture<byte[]> get(String key) {
    return simpleStore.get(key);
  }

  @Override
  public ListenableFuture<byte[]> put(String key, @Nullable byte[] value) {
    return simpleStore.put(key, value);
  }

  @Override
  public ListenableFuture<Boolean> contains(String key) {
    return simpleStore.contains(key);
  }

  @Override
  public ListenableFuture<Void> deleteAll() {
    return simpleStore.deleteAll();
  }

  @Override
  public void close() {
    simpleStore.close();
  }

  @Override
  public ListenableFuture<Integer> getInt(String key) {
    return Futures.transform(
        get(key),
        (b) -> {
          if (b == null || b.length != 4) {
            return 0;
          }
          // decode big endian
          return b[0] << 24 | (b[1] & 0xFF) << 16 | (b[2] & 0xFF) << 8 | (b[3] & 0xFF);
        },
        directExecutor());
  }

  @Override
  public ListenableFuture<Integer> put(String key, int value) {
    byte[] bytes;
    if (value != 0) {
      // encode big endian
      bytes =
          new byte[] {
            (byte) (value >> 24), (byte) (value >> 16), (byte) (value >> 8), (byte) value
          };
    } else {
      bytes = null;
    }
    return Futures.transform(put(key, bytes), (v) -> value, directExecutor());
  }

  @Override
  public ListenableFuture<Long> getLong(String key) {
    return Futures.transform(
        get(key),
        (b) -> {
          if (b == null || b.length != 8) {
            return 0L;
          }
          return (b[0] & 0xFFL) << 56
              | (b[1] & 0xFFL) << 48
              | (b[2] & 0xFFL) << 40
              | (b[3] & 0xFFL) << 32
              | (b[4] & 0xFFL) << 24
              | (b[5] & 0xFFL) << 16
              | (b[6] & 0xFFL) << 8
              | (b[7] & 0xFFL);
        },
        directExecutor());
  }

  @Override
  public ListenableFuture<Long> put(String key, long value) {
    byte[] bytes;
    if (value != 0) {
      long v = value;
      bytes = new byte[8];
      // encode big endian
      for (int i = 7; i >= 0; i--) {
        bytes[i] = (byte) (v & 0xffL);
        v >>= 8;
      }
    } else {
      bytes = null;
    }
    return Futures.transform(put(key, bytes), (v) -> value, directExecutor());
  }

  @Override
  public ListenableFuture<Boolean> getBoolean(String key) {
    return Futures.transform(
        get(key), (b) -> b != null && b.length > 0 && b[0] > 0, directExecutor());
  }

  @Override
  public ListenableFuture<Boolean> put(String key, boolean value) {
    byte[] bytes;
    if (value) {
      bytes = new byte[] {1};
    } else {
      bytes = new byte[] {0};
    }
    return Futures.transform(put(key, bytes), (v) -> value, directExecutor());
  }

  @Override
  public ListenableFuture<Double> getDouble(String key) {
    return Futures.transform(getLong(key), Double::longBitsToDouble, directExecutor());
  }

  @Override
  public ListenableFuture<Double> put(String key, double value) {
    return Futures.transform(
        put(key, Double.doubleToRawLongBits(value)), (v) -> value, directExecutor());
  }

  @Override
  public ListenableFuture<Void> remove(String key) {
    return simpleStore.remove(key);
  }
}

File: simplestore/src/main/java/com/uber/simplestore/primitive/PrimitiveSimpleStoreImpl.kt
/*
 * Copyright (C) 2019. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.simplestore.primitive

import com.uber.simplestore.SimpleStore

internal class PrimitiveSimpleStoreImpl(
    private val simpleStore: SimpleStore
) : PrimitiveSimpleStore {

    override suspend fun getString(key: String): String {
        val value = simpleStore.getString(key)
        return value.ifEmpty { "" }
    }

    override suspend fun put(key: String, value: String): String {
        return simpleStore.putString(key, value)
    }

    override suspend fun putString(key: String, value: String?): String {
        return simpleStore.putString(key, value)
    }

    override suspend fun get(key: String): ByteArray {
        return simpleStore.get(key)
    }

    override suspend fun put(key: String, value: ByteArray?): ByteArray {
        return simpleStore.put(key, value)
    }

    override suspend fun contains(key: String): Boolean {
        return simpleStore.contains(key)
    }

    override suspend fun clear() {
        simpleStore.clear()
    }

    override suspend fun deleteAllNow() {
        simpleStore.deleteAllNow()
    }

    override fun close() {
        simpleStore.close()
    }

    override suspend fun getInt(key: String): Int {
        val bytes = get(key)
        return when {
            bytes.isEmpty() || bytes.size != 4 -> 0
            else -> {
                // decode big endian
                bytes[0].toInt() shl 24 or
                        (bytes[1].toInt() and 0xFF) shl 16 or
                        (bytes[2].toInt() and 0xFF) shl 8 or
                        (bytes[3].toInt() and 0xFF)
            }
        }
    }

    override suspend fun put(key: String, value: Int): Int {
        val bytes = if (value != 0) {
            // encode big endian
            byteArrayOf(
                (value shr 24).toByte(),
                (value shr 16).toByte(),
                (value shr 8).toByte(),
                value.toByte()
            )
        } else {
            null
        }
        put(key, bytes)
        return value
    }

    override suspend fun getLong(key: String): Long {
        val bytes = get(key)
        return when {
            bytes.isEmpty() || bytes.size != 8 -> 0L
            else -> {
                (bytes[0].toLong() and 0xFF) shl 56 or
                        (bytes[1].toLong() and 0xFF) shl 48 or
                        (bytes[2].toLong() and 0xFF) shl 40 or
                        (bytes[3].toLong() and 0xFF) shl 32 or
                        (bytes[4].toLong() and 0xFF) shl 24 or
                        (bytes[5].toLong() and 0xFF) shl 16 or
                        (bytes[6].toLong() and 0xFF) shl 8 or
                        (bytes[7].toLong() and 0xFF)
            }
        }
    }

    override suspend fun put(key: String, value: Long): Long {
        val bytes = if (value != 0L) {
            val bytes = ByteArray(8)
            var v = value
            // encode big endian
            for (i in 7 downTo 0) {
                bytes[i] = (v and 0xffL).toByte()
                v = v shr 8
            }
            bytes
        } else {
            null
        }
        put(key, bytes)
        return value
    }

    override suspend fun getBoolean(key: String): Boolean {
        val bytes = get(key)
        return bytes.isNotEmpty() && bytes[0] > 0
    }

    override suspend fun put(key: String, value: Boolean): Boolean {
        val bytes = if (value) {
            byteArrayOf(1)
        } else {
            byteArrayOf(0)
        }
        put(key, bytes)
        return value
    }

    override suspend fun getDouble(key: String): Double {
        val longValue = getLong(key)
        return Double.fromBits(longValue)
    }

    override suspend fun put(key: String, value: Double): Double {
        put(key, value.toBits())
        return value
    }

    override suspend fun remove(key: String) {
        simpleStore.remove(key)
    }
}